src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,336 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "piecesbar.h"
+
+#include <QApplication>
+#include <QHelpEvent>
+#include <QPainter>
+#include <QTextStream>
+#include <QToolTip>
+
+#include "base/bittorrent/torrenthandle.h"
+#include "base/utils/misc.h"
+
+#define TYPESET_FILELIST_AS_TABLE // TODO remove this #ifdef before merging to master
+
+namespace
+{
+    class PieceIndexToImagePos
+    {
+    public:
+        PieceIndexToImagePos(const BitTorrent::TorrentHandle*const torrent, const QImage &image)
+            : m_piecesCount {torrent->info().piecesCount()}
+            , m_imageWidth {image.width()}
+        {
+        }
+
+        int imagePos(int pieceIndex) const
+        {
+            return m_piecesCount == 1 ? 0
+                   : static_cast<int>(static_cast<float>(pieceIndex) / (m_piecesCount - 1) * (m_imageWidth - 1));
+        }
+
+        int pieceIndex(int imagePos) const
+        {
+            return m_imageWidth == 1 ? 0
+                   : static_cast<int>(static_cast<float>(imagePos) / (m_imageWidth - 1) * (m_piecesCount - 1));
+        }
+
+    private:
+        const int m_piecesCount;
+        const int m_imageWidth;
+    };
+
+    class DetailedTooltipRenderer
+    {
+    public:
+        DetailedTooltipRenderer(QTextStream &stream, const QString &header)
+            : m_stream(stream)
+        {
+            m_stream << header
+#ifdef TYPESET_FILELIST_AS_TABLE
+                << R"(<table style="width:100%; padding: 3px; vertical-align: middle;">)";
+#else
+                << "<ul>";
+#endif
+        }
+
+        ~DetailedTooltipRenderer()
+        {
+#ifdef TYPESET_FILELIST_AS_TABLE
+            m_stream << "</table>";
+#else
+            m_stream << "</ul>";
+#endif
+        }
+
+        void operator()(const QString &size, const QString &path)
+        {
+#ifdef TYPESET_FILELIST_AS_TABLE
+            m_stream << R"(<tr><td style="white-space:nowrap">)" << size << "</td><td>" << path << "</td></tr>";
+#else
+            m_stream << "<li>[" << size << "] " << path << "</li>";
+#endif
+        }
+
+    private:
+        QTextStream &m_stream;
+    };
+}
+
+PiecesBar::PiecesBar(QWidget *parent)
+    : QWidget{parent}
```

Comment:
And please don't use uncrustify now unless you sure it will not produce another coding style violations. But you can use discovered violations to upgrade uncrustify.cfg later.